types/react-redux/react-redux-tests.tsx

```diff
@@ -1601,3 +1598,26 @@ function testPreserveDiscriminatedUnions() {
     <ConnectedMyText type="localized" color="red" />; // $ExpectError
     <ConnectedMyText type="localized" color="red" params={someParams} />;
 }
+
+/**
+ * The `withStyles` HOC in MUI returns a `JSXElementConstructor` instead of a `ComponentType` as of
+ * https://github.com/mui/material-ui/pull/24746.
+ */
+function testJSXElementConstructorIsAllowed() {
```

Comment:
This test is working fine with existing typings: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeuEAdm1vgL5xEnlKeGAFpKAEwCuAD1o0A9AgQ1EcACoALLHAAGAd2AwNAZRgBPADZZUOuAAkA8gGE4wNnACyAVQCScSjCSUGzoyLoAUsYAGgCiViBYbDBO7KgwUJL40LZuaTjicBAEKLop4OyJMGpmYFi2aIUEKkgaMDBgqABc8vIA5oYakgBGAHSsIPIgksCTyDBYUMDIFiLT8mCSFhbyAEwALADsewBsI83yNASSbPjA7HDzaZGx8ZUpIemZMNA+qACCWwgeiw4gAFABKZgqOAwtzzKAEPDaAAKxA6UJhmMxBAgEE6cDYkhAQwWAG5oViYUNkFB8UNcVZkGxyVjuDQKXBWB91NZkiRIJwkvjsMIRs84lgEkl3mkMlkoAAeVEQDoAPjgAF44BDNerCVtyRy4QtEbhtKZoFhTHNtEwOVi0jbNG5el0CUSSVAWZi2ViuWk4CBkGBrfM1BBleitaDHfN8RbKKGsJCNerQXbKdjcfjY1gRrnnWxXQAaDnccGGv2peDvTj4EFlAWVTWc9h1mCgoMhmBOiNo1Dg0FqXmNipJCschW1rjzcSjwXwalQODyVXk7hAA

You may have a slightly outdated version of `@types/react` installed. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56789 or https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56413 probably fixed potential problems with `JSXElementConstructor`

Reply:
Thanks for taking a look, you're right my test was too reductive and worked with the old types. I believe I have the latest types installed for everything (i.e., `@types/react@17.0.39`).

I updated the test so that it now fails with the old types ([Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeuEAdm1vgL5xEnlKeGAFpKAEwCuAD1p02MLFAJ4scAArEw6JjTj6+ECAC44bSSABGSgNx6Dl5FFOWjAGxxs73GjXFc3JzVWNlR4ABUsMIBhEkhOBVNsYQA6ACkAZQANAFEPECwFWNCYKEl8aAAeTQhtAD47GgB6BAQ9JHCACzUAAwB3YBhOjJgATw9UHrgACQB5aLhgNjgAWQBVAEk4ShhJKFCUOB7M3PzCmGKwsoqoKaWwnHE4CAJDntjwdnPw0bAsKbQzwI7TgnRgMG0xiaTQA5oNOpJLClWCAmiBJMA0chFFBgMg3CIMU0wJI3G4mgAmAAsAHYqQA2FLtJo0AiSNj4YDsOCKMInPJYApFdhXcowaAbVAAQTJED6WHEAAoAJTMXwGRYKJQqXBqEbQLAjbFqXQajVhY1dJYw1CmcxWWz2fQ+DUhMJwEDIMBGxThCA1bRwAC8cEVFsUpn1lB9WFVQbqodNZv0BCMpnDWBSGatbBtABonXBuMrGq6RfBipx8AqPvFzsG4CEqzBFZ7vTBLf6tKhlYrIjE4l8FCXC5VK1xFOJa0P4I4oHAmg0aD4gA)):
```diff
- const TestComponent: React.JSXElementConstructor<Props> = () => null;
+ declare const TestComponent: React.JSXElementConstructor<Props>;
```

I think it was narrowing the type to only be the function type and not considering the class type since I'm also able to make it fail by assigning a class to it instead of a function.

I imagine this PR might not be the best way to fix this issue, but I'm not familiar enough with all the moving pieces to know where to look for the root issue. Can you point me in the right direction?